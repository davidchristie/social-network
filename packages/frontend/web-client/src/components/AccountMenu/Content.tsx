import { Avatar, Menu } from "design-system";
import React from "react";
import ButtonLink from "../ButtonLink";
import LogoutButton from "../LogoutButton";

interface Account {
  profile: {
    avatar: {
      url: string;
    } | null;
    id: string;
  };
}

export interface Props {
  account: Account;
}

interface State {
  anchorElement: HTMLElement | null;
}

export default class Content extends React.Component<Props, State> {
  public state = {
    anchorElement: null,
  };

  public render () {
    const { account } = this.props;
    return (
      <div>
        {this.renderAvatar(account)}
        {this.renderMenu(account)}
      </div>
    );
  }

  private renderAccountButton = () => {
    return (
      <ButtonLink
        onClick={this.handleClose}
        to="/account"
      >
        Account
      </ButtonLink>
    );
  }

  private renderProfileButton = (account: Account) => {
    return (
      <ButtonLink
        onClick={this.handleClose}
        to={`profile/${account.profile.id}`}
      >
        Profile
      </ButtonLink>
    );
  }

  private handleClick = event => {
    console.log(event.currentTarget);
    this.setState({
      anchorElement: event.currentTarget,
    });
  }

  private handleClose = () => {
    this.setState({
      anchorElement: null,
    });
  }

  private renderAvatar = (account: Account) => {
    const { profile: { avatar } } = account;
    const image = avatar ? avatar.url : undefined;
    return (
      <Avatar
        image={image}
        onClick={this.handleClick}
        size="small"
      />
    );
  }

  private renderMenu = (account: Account) => {
    return (
      <Menu
        anchorElement={this.state.anchorElement}
        onClose={this.handleClose}
        open={Boolean(this.state.anchorElement)}
      >
        {this.renderProfileButton(account)}
        {this.renderAccountButton()}
        <hr />
        <LogoutButton />
      </Menu>
    );
  }
}
