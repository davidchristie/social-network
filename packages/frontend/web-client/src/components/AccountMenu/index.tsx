import { Alert, Avatar, Loading } from "design-system";
import React from "react";

import { ImageData } from "../../fragments/Image";
import AccountQuery, { Result } from "../AccountQuery";
import ButtonLink from "../ButtonLink";
import Dropdown from "../Dropdown";
import LogoutButton from "../LogoutButton";
import "./index.css";

interface State {
  isDropdownOpen: boolean;
}

export class AccountMenuContent extends React.Component<Result, State> {
  public state = {
    isDropdownOpen: false,
  };

  public render () {
    const { data, error, loading } = this.props;
    if (loading || !data || !data.account) {
      return <Loading />;
    }
    if (error) {
      return <Alert>{error.message}</Alert>;
    }
    const { account } = data;
    return (
      <div className="AccountMenu">
        {this.renderAvatar(account)}
        {this.renderDropdown(account)}
      </div>
    );
  }

  private closeDropdown = () => {
    this.setState({
      isDropdownOpen: false,
    });
  }

  private renderAccountButton = () => {
    return (
      <ButtonLink
        onClick={this.closeDropdown}
        to="/account"
      >
        Account
      </ButtonLink>
    );
  }

  private renderAvatar = (
    account: { profile: { avatar: ImageData | null } }
  ) => {
    const { profile: { avatar } } = account;
    const image = avatar ? avatar.url : undefined;
    return (
      <Avatar
        image={image}
        onClick={this.openDropdown}
        size="small"
      />
    );
  }

  private renderDropdown = (
    account: { profile: { id: string } }
  ) => {
    return (
      <Dropdown
        onClose={this.closeDropdown}
        open={this.state.isDropdownOpen}
      >
        {this.renderProfileButton(account)}
        {this.renderAccountButton()}
        <hr />
        <LogoutButton />
      </Dropdown>
    );
  }

  private renderProfileButton = (
    account: { profile: { id: string } }
  ) => {
    return (
      <ButtonLink
        onClick={this.closeDropdown}
        to={`profile/${account.profile.id}`}
      >
        Profile
      </ButtonLink>
    );
  }

  private openDropdown = () => {
    this.setState({
      isDropdownOpen: true,
    });
  }
}

const AccountMenu = () => (
  <AccountQuery>
    {result => {
      return <AccountMenuContent {...result} />;
    }}
  </AccountQuery>
);

export default AccountMenu;
