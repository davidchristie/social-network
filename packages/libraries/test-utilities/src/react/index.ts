export { default as describeMountedElement } from "./describeMountedElement";
export { default as describeWithCustomClass } from "./describeWithCustomClass";
export { default as describeWithNoProps } from "./describeWithNoProps";
export { default as describeWithProps } from "./describeWithProps";
export { default as itMatchesSnapshot } from "./itMatchesSnapshot";
