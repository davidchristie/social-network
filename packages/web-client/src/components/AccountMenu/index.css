.AccountMenu {
  position: relative;
}

.AccountMenu .avatar {
  background: white;
  border-radius: 50%;
  height: 35px;
  width: 35px;
}
