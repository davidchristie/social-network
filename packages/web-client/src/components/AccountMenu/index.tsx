import React from "react";

import { ImageData } from "../../fragments/Image";
import AccountQuery from "../AccountQuery";
import Alert from "../Alert";
import Avatar from "../Avatar";
import ButtonLink from "../ButtonLink";
import Dropdown from "../Dropdown";
import LogoutButton from "../LogoutButton";
import "./index.css";

interface State {
  isDropdownOpen: boolean;
}

export default class AccountMenu extends React.Component<{}, State> {
  public state = {
    isDropdownOpen: false,
  };

  public render () {
    return (
      <AccountQuery>
        {({ data, error, loading }) => {
          if (loading || !data || !data.account) {
            return "Loading";
          }
          if (error) {
            return <Alert>{error.message}</Alert>;
          }
          const { account } = data;
          return (
            <div className="AccountMenu">
              {this.renderAvatar(account)}
              {this.renderDropdown(account)}
            </div>
          );
        }}
      </AccountQuery>
    );
  }

  private closeDropdown = () => {
    this.setState({
      isDropdownOpen: false,
    });
  }

  private renderAccountButton = () => {
    return (
      <ButtonLink
        onClick={this.closeDropdown}
        to="/account"
      >
        Account
      </ButtonLink>
    );
  }

  private renderAvatar = (
    account: { profile: { avatar: ImageData | null } }
  ) => {
    const { profile: { avatar } } = account;
    const image = avatar ? avatar.url : undefined;
    return (
      <Avatar
        image={image}
        onClick={this.openDropdown}
        size="small"
      />
    );
  }

  private renderDropdown = (
    account: { profile: { id: string } }
  ) => {
    return (
      <Dropdown
        onClose={this.closeDropdown}
        open={this.state.isDropdownOpen}
      >
        {this.renderProfileButton(account)}
        {this.renderAccountButton()}
        <hr />
        <LogoutButton />
      </Dropdown>
    );
  }

  private renderProfileButton = (
    account: { profile: { id: string } }
  ) => {
    return (
      <ButtonLink
        onClick={this.closeDropdown}
        to={`profile/${account.profile.id}`}
      >
        Profile
      </ButtonLink>
    );
  }

  private openDropdown = () => {
    this.setState({
      isDropdownOpen: true,
    });
  }
}
