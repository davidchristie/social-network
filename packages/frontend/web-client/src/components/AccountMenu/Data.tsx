import { Alert, Loading } from "design-system";
import React from "react";
import AccountQuery from "../AccountQuery";
import { Props as ContentProps } from "./Content";

interface Props {
  content: React.ComponentType<ContentProps>;
}

const Data: React.ComponentType<Props> = ({ content }) => {
  const Content = content;
  return (
    <AccountQuery>
      {({ error, loading, data }) => {
        if (error) {
          return <Alert>{error.message}</Alert>;
        }
        if (loading || !data || !data.account) {
          return <Loading />;
        }
        return <Content account={data.account} />;
      }}
    </AccountQuery>
  );
};

export default Data;
