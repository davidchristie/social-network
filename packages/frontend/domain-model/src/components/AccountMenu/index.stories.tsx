import { withInfo } from "@storybook/addon-info";
import { storiesOf } from "@storybook/react";
import apolloStorybookDecorator from "apollo-storybook-react";
import { Theme } from "design-system";
import "material-design-icons-iconfont/dist/material-design-icons.css";
import React from "react";
import { MemoryRouter } from "react-router-dom";
import AccountMenu from ".";

const typeDefs = `
  schema {
    query: Query
  }

  type Account {
    email: String!
    id: ID!
    name: String!
    profile: Profile!
  }

  type Image {
    id: ID!
    url: String!
  }

  type Profile {
    avatar: Image
    id: ID!
    followers: [Profile!]!
    following: [Profile!]!
    name: String!
  }

  type Query {
    account: Account
  }
`;

const mocks = {
  Query: () => {
    return {
      account: () => ({
        email: "account@email.com",
        id: "account_id",
        name: "Account",
        profile: () => ({
          avatar: () => null,
          followers: () => ([]),
          following: () => ([]),
          id: "profile_id",
          name: "Profile",
        }),
      }),
    };
  },
};

const stories = storiesOf("Domain Model", module);
stories.addDecorator(getStory => <MemoryRouter>{getStory()}</MemoryRouter>);
stories.addDecorator(
  apolloStorybookDecorator({
    mocks,
    typeDefs,
  }),
);
stories.addDecorator(getStory => <Theme>{getStory()}</Theme>);
stories.addDecorator(withInfo({
  inline: true,
  propTablesExclude: [
    MemoryRouter,
    function StorybookProvider () { return null; },
    Theme,
  ],
}));
stories.add(
  "AccountMenu",
  () => <AccountMenu />
);
