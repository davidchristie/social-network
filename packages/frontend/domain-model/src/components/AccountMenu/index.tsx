import React from "react";
import Content from "./Content";
import Data from "./Data";
import "./index.css";

const AccountMenu: React.ComponentType<{}> = () => <Data content={Content} />;

export default AccountMenu;
