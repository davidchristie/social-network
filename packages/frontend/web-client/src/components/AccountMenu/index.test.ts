import { itRendersWithoutCrashing } from "test-utilities/dist/react";
import AccountMenu from ".";

describe("AccountMenu component", () => {
  itRendersWithoutCrashing(AccountMenu);
});
