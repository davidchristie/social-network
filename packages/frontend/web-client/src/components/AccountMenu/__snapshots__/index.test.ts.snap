// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountMenu component with no props matches snapshot 1`] = `
<AccountQuery>
  <Component />
</AccountQuery>
`;

exports[`AccountMenuContent component renders loading state matches snapshot 1`] = `<Loading />`;

exports[`AccountMenuContent component renders profile with no avatar matches snapshot 1`] = `
<div
  className="AccountMenu"
>
  <Avatar
    onClick={[Function]}
    size="small"
  />
  <Dropdown
    onClose={[Function]}
    open={false}
  >
    <ButtonLink
      onClick={[Function]}
      to="profile/PROFILE_ID"
    >
      Profile
    </ButtonLink>
    <ButtonLink
      onClick={[Function]}
      to="/account"
    >
      Account
    </ButtonLink>
    <hr />
    <LogoutButton />
  </Dropdown>
</div>
`;
